src/analytics/api/models/events.py

```diff
@@ -266,6 +268,58 @@ def get_d3_chart_events(self, sites, start_date, end_date, pollutant, frequency)
                 .exec()
         )
 
+    @cache.memoize()
+    def get_d3_chart_events_v2(self, sites, start_date, end_date, pollutant, frequency, tenant):
+
+        if pollutant not in ["pm2_5", "pm10", "no2", "pm1"]:
+            raise Exception("Invalid pollutant")
+
+        columns = ["site_id", "name", "timestamp as time", "description as generated_name", f"{pollutant} as value"]
+
+        query = f"""
+          SELECT {', '.join(map(str, columns))} 
+          FROM {self.BIGQUERY_EVENTS}
+          JOIN {self.BIGQUERY_SITES} ON {self.BIGQUERY_SITES}.id = {self.BIGQUERY_EVENTS}.site_id 
+          WHERE  {self.BIGQUERY_EVENTS}.timestamp >= '{start_date}'
+          AND {self.BIGQUERY_EVENTS}.timestamp <= '{end_date}'
+          AND {self.BIGQUERY_EVENTS}.tenant = '{tenant}'
+          AND `airqo-250220.metadata.sites`.id in UNNEST({sites})
+        """
+
+        client = bigquery.Client()
+        job_config = bigquery.QueryJobConfig()
+        job_config.use_query_cache = True
+
+        dataframe = client.query(query, job_config).result().to_dataframe()
+        dataframe["value"] = dataframe["value"].apply(lambda x: round(x, 2))
+        site_groups = dataframe.groupby("site_id")
+
+        data = []
+        if frequency.lower() == "daily":
+            resample_value = "24H"
+        elif frequency.lower() == "monthly":
+            resample_value = "720H"
+        elif frequency.lower() == "hourly":
+            resample_value = "1H"
+        else:
+            resample_value = "1H"
+
+        for _, site_group in site_groups:
+
+            values = site_group[["value", "time"]]
+            ave_values = pd.DataFrame(values.resample(resample_value, on="time").mean())
+            ave_values["time"] = ave_values.index
+            ave_values["time"] = ave_values["time"].apply(date_to_str)
+            ave_values = ave_values.reset_index(drop=True)
```

Comment:
Averaging the data based on the frequency

Reply:
Thanks @NoahNsimbe , so I gather that we might do some of this averaging from the pipeline....in the near future. Right?